<a href="https://colab.research.google.com/github/NataliProhorova/Data-_Pyton-_Code_Info/blob/main/Lottery.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Find a most potentially winner numbers and 2 lucky stars   combination

In [ ]:
# prompt: df info

df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   DrawDate                    52 non-null     object
 1   Ball 1                      52 non-null     int64 
 2   Ball 2                      52 non-null     int64 
 3   Ball 3                      52 non-null     int64 
 4   Ball 4                      52 non-null     int64 
 5   Ball 5                      52 non-null     int64 
 6   Lucky Star 1                52 non-null     int64 
 7   Lucky Star 2                52 non-null     int64 
 8   UK Millionaire Maker        52 non-null     object
 9   European Millionaire Maker  1 non-null      object
 10  DrawNumber                  52 non-null     int64 
dtypes: int64(8), object(3)
memory usage: 4.6+ KB


In [ ]:
# prompt: import pandas

import pandas as pd


In [ ]:
df=pd.read_csv('/content/euromillions-draw-history.csv')

In [ ]:
df.head()

DrawDate  Ball 1  Ball 2  Ball 3  Ball 4  Ball 5  Lucky Star 1  \
0  04-Apr-2025      27      39      41      45      50             5   
1  01-Apr-2025       9      24      30      46      49             4   
2  28-Mar-2025      10      21      30      42      45             1   
3  25-Mar-2025       1       3      18      33      36             2   
4  21-Mar-2025       9      17      18      21      50             5   

   Lucky Star 2 UK Millionaire Maker European Millionaire Maker  DrawNumber  
0             7            TNKB70417                        NaN        1830  
1            12            ZMJR93122                        NaN        1829  
2             9            HLHV99083                        NaN        1828  
3             6            MKHJ22564                        NaN        1827  
4            12            MJFW44014                        NaN        1826

In [ ]:
df.tail()

DrawDate  Ball 1  Ball 2  Ball 3  Ball 4  Ball 5  Lucky Star 1  \
47  22-Oct-2024       3      13      27      30      46             1   
48  18-Oct-2024       4      15      23      30      39             1   
49  15-Oct-2024       2      15      32      36      48             3   
50  11-Oct-2024      16      23      32      46      49             4   
51  08-Oct-2024       7      14      19      34      40             6   

    Lucky Star 2 UK Millionaire Maker European Millionaire Maker  DrawNumber  
47             2            ZFDM17741                        NaN        1783  
48            12            JDDF38095                        NaN        1782  
49             9            ZCCN33395                        NaN        1781  
50             5            MZBX76775                        NaN        1780  
51             8            ZZBZ48011                        NaN        1779

In [ ]:
# Assuming you're using pandas
# Replace 'colum' with the actual column name and 'df' with your DataFrame variable
df = df.drop(columns=['UK Millionaire Maker'])

In [ ]:
from collections import Counter
 # Most frequent numbers in Lucky Star 1
ls1_counts = Counter(df['Lucky Star 1'])
 # Most frequent numbers in Lucky Star 2
ls2_counts = Counter(df['Lucky Star 2'])
# Most frequent combinations of (Lucky Star 1, Lucky Star 2)
from collections import Counter
 # Most frequent numbers in Lucky Star 1
ls1_counts = Counter(df['Lucky Star 1'])
 # Most frequent numbers in Lucky Star 2
ls2_counts = Counter(df['Lucky Star 2'])
# Most frequent combinations of (Lucky Star 1, Lucky Star 2)
lucky_star_combinations = list(zip(df['Lucky Star 1'], df['Lucky Star 2']))
combination_counts = Counter(lucky_star_combinations)
#return ls1_counts, ls2_counts, combination_counts
print(ls1_counts, ls2_counts, combination_counts)

Counter({5: 8, 1: 8, 2: 8, 6: 8, 4: 6, 8: 5, 3: 4, 7: 2, 11: 1, 9: 1, 10: 1}) Counter({9: 13, 12: 9, 7: 6, 8: 6, 10: 6, 11: 5, 6: 3, 5: 2, 3: 1, 2: 1}) Counter({(5, 9): 4, (1, 9): 3, (2, 7): 3, (6, 8): 3, (5, 7): 2, (4, 12): 2, (2, 6): 2, (8, 11): 2, (2, 9): 2, (3, 8): 2, (4, 5): 2, (6, 9): 2, (5, 12): 1, (7, 11): 1, (6, 11): 1, (2, 11): 1, (3, 10): 1, (1, 6): 1, (8, 10): 1, (5, 10): 1, (6, 7): 1, (8, 12): 1, (4, 8): 1, (11, 12): 1, (4, 10): 1, (6, 10): 1, (9, 12): 1, (7, 12): 1, (1, 10): 1, (1, 3): 1, (10, 12): 1, (8, 9): 1, (1, 2): 1, (1, 12): 1, (3, 9): 1})


In [ ]:
lucky_star_combinations = list(zip(df['Lucky Star 1'], df['Lucky Star 2']))
combination_counts = Counter(lucky_star_combinations)
print("Lucky Star Combination Frequency (Order Matters):", combination_counts.most_common())

Lucky Star Combination Frequency (Order Matters): [((5, 9), 4), ((1, 9), 3), ((2, 7), 3), ((6, 8), 3), ((5, 7), 2), ((4, 12), 2), ((2, 6), 2), ((8, 11), 2), ((2, 9), 2), ((3, 8), 2), ((4, 5), 2), ((6, 9), 2), ((5, 12), 1), ((7, 11), 1), ((6, 11), 1), ((2, 11), 1), ((3, 10), 1), ((1, 6), 1), ((8, 10), 1), ((5, 10), 1), ((6, 7), 1), ((8, 12), 1), ((4, 8), 1), ((11, 12), 1), ((4, 10), 1), ((6, 10), 1), ((9, 12), 1), ((7, 12), 1), ((1, 10), 1), ((1, 3), 1), ((10, 12), 1), ((8, 9), 1), ((1, 2), 1), ((1, 12), 1), ((3, 9), 1)]


In [ ]:
ls1_counts = Counter(df['Lucky Star 1'])
print("Lucky Star 1 Frequency:", ls1_counts.most_common())

Lucky Star 1 Frequency: [(5, 8), (1, 8), (2, 8), (6, 8), (4, 6), (8, 5), (3, 4), (7, 2), (11, 1), (9, 1), (10, 1)]


This output shows the frequency of each Lucky Star number that appeared in the 'Lucky Star 1' column of your EuroMillions draw history data across the entire dataset.

Interpretation:

The list contains tuples, where each tuple is (Lucky Star Number, Frequency). The list is ordered by frequency, with the most frequent numbers appearing first.

(5, 8): The number 5 appeared 8 times in the 'Lucky Star 1' column.
(1, 8): The number 1 appeared 8 times in the 'Lucky Star 1' column.
(2, 8): The number 2 appeared 8 times in the 'Lucky Star 1' column.
(6, 8): The number 6 appeared 8 times in the 'Lucky Star 1' column.
(4, 6): The number 4 appeared 6 times in the 'Lucky Star 1' column.
(8, 5): The number 8 appeared 5 times in the 'Lucky Star 1' column.
(3, 4): The number 3 appeared 4 times in the 'Lucky Star 1' column.
(7, 2): The number 7 appeared 2 times in the 'Lucky Star 1' column.
(11, 1): The number 11 appeared 1 time in the 'Lucky Star 1' column.
(9, 1): The number 9 appeared 1 time in the 'Lucky Star 1' column.
(10, 1): The number 10 appeared 1 time in the 'Lucky Star 1' column.
Key Takeaway:


Based on your entire draw history, the Lucky Star numbers 1, 2, 5, and 6 have appeared most frequently in the 'Lucky Star 1' column. Numbers 9, 10, and 11 have appeared the least frequently in this column.

In [ ]:
ls2_counts = Counter(df['Lucky Star 2'])
print("Lucky Star 2 Frequency:", ls2_counts.most_common())

Lucky Star 2 Frequency: [(9, 13), (12, 9), (7, 6), (8, 6), (10, 6), (11, 5), (6, 3), (5, 2), (3, 1), (2, 1)]


# This output shows the frequency of each Lucky Star number that appeared in the 'Lucky Star 2' column of your EuroMillions draw history data across the entire dataset.

Interpretation:

Similar to the 'Lucky Star 1' frequency, this list contains tuples of (Lucky Star Number, Frequency), ordered from most frequent to least frequent.

(9, 13): The number 9 appeared 13 times in the 'Lucky Star 2' column.
(12, 9): The number 12 appeared 9 times in the 'Lucky Star 2' column.
(7, 6): The number 7 appeared 6 times in the 'Lucky Star 2' column.
(8, 6): The number 8 appeared 6 times in the 'Lucky Star 2' column.
(10, 6): The number 10 appeared 6 times in the 'Lucky Star 2' column.
(11, 5): The number 11 appeared 5 times in the 'Lucky Star 2' column.
(6, 3): The number 6 appeared 3 times in the 'Lucky Star 2' column.
(5, 2): The number 5 appeared 2 times in the 'Lucky Star 2' column.
(3, 1): The number 3 appeared 1 time in the 'Lucky Star 2' column.
(2, 1): The number 2 appeared 1 time in the 'Lucky Star 2' column.
Key Takeaway:

Across your entire draw history, the Lucky Star number 9 has been the most frequent in the 'Lucky Star 2' column, appearing significantly more often than other numbers in this column. Numbers 2 and 3 have appeared the least frequently in 'Lucky Star 2'.

# Expected appearances of numbers 1, 2, 5, or 6 in 'Lucky Star 1' in 100 draws: Each would have an expected value of approximately 9.09 times.


In [ ]:
import pandas as pd
from scipy.stats import binom

def probability_specific_lucky_star_frequency(df, number, column, n_future_draws, historical_analysis_length=None):
    """
    Calculates the probability of a specific Lucky Star number appearing a
    certain number of times in a given number of future draws, based on its
    theoretical probability.

    Args:
        df (pd.DataFrame): The DataFrame containing the draw data.
        number (int): The Lucky Star number to analyze.
        column (str): The Lucky Star column ('Lucky Star 1' or 'Lucky Star 2').
        n_future_draws (int): The number of future draws to consider.
        historical_analysis_length (int, optional): If specified, use the
                                                     frequency from the last
                                                     this many draws to
                                                     estimate probability.
                                                     Defaults to None (uses
                                                     theoretical probability).

    Returns:
        dict: A dictionary containing the probability of the number appearing
              k times (where k ranges from 0 to n_future_draws).
    """
    if column not in ['Lucky Star 1', 'Lucky Star 2']:
        print("Error: Invalid column specified. Use 'Lucky Star 1' or 'Lucky Star 2'.")
        return {}

    if column == 'Lucky Star 1':
        n_possible = 11
    else:
        n_possible = 12

    if historical_analysis_length is not None:
        df_historical = df.tail(historical_analysis_length)
        historical_count = df_historical[column][df_historical[column] == number].count()
        probability_of_occurrence = historical_count / historical_analysis_length
    else:
        probability_of_occurrence = 1 / n_possible

    probabilities = {}
    for k in range(n_future_draws + 1):
        probability = binom.pmf(k, n_future_draws, probability_of_occurrence)
        probabilities[k] = probability

    return probabilities

if __name__ == "__main__":
    # Load your EuroMillions draw history data
    df = pd.read_csv('/content/euromillions-draw-history.csv')

    n_future_draws = 100
    analyzed_number_ls1 = 1  # Example number for Lucky Star 1
    analyzed_number_ls2 = 9  # Example number for Lucky Star 2

    print(f"--- Probability of Specific Frequencies in Next {n_future_draws} Draws ---")

    # Probability for Lucky Star 1 (theoretical)
    prob_ls1_theoretical = probability_specific_lucky_star_frequency(
        df, analyzed_number_ls1, 'Lucky Star 1', n_future_draws
    )
    print(f"\nLucky Star 1 - Number {analyzed_number_ls1} (Theoretical Probability = 1/11):")
    print(f"Probability of appearing exactly 9 times: {prob_ls1_theoretical.get(9, 0):.4f}")
    print(f"Probability of appearing exactly 10 times: {prob_ls1_theoretical.get(10, 0):.4f}")



--- Probability of Specific Frequencies in Next 100 Draws ---

Lucky Star 1 - Number 1 (Theoretical Probability = 1/11):
Probability of appearing exactly 9 times: 0.1380
Probability of appearing exactly 10 times: 0.1256


In [ ]:
import pandas as pd
from scipy.stats import binom

def probability_specific_lucky_star_frequency(df, number, column, n_future_draws, historical_analysis_length=None):
    """
    Calculates the probability of a specific Lucky Star number appearing a
    certain number of times in a given number of future draws, based on its
    theoretical probability or historical frequency.

    Args:
        df (pd.DataFrame): The DataFrame containing the draw data.
        number (int): The Lucky Star number to analyze.
        column (str): The Lucky Star column ('Lucky Star 1' or 'Lucky Star 2').
        n_future_draws (int): The number of future draws to consider.
        historical_analysis_length (int, optional): If specified, use the
                                                     frequency from the last
                                                     this many draws to
                                                     estimate probability.
                                                     Defaults to None (uses
                                                     theoretical probability).

    Returns:
        dict: A dictionary containing the probability of the number appearing
              k times (where k ranges from 0 to n_future_draws).
    """
    if column not in ['Lucky Star 1', 'Lucky Star 2']:
        print("Error: Invalid column specified. Use 'Lucky Star 1' or 'Lucky Star 2'.")
        return {}

    if column == 'Lucky Star 1':
        n_possible = 11
    else:
        n_possible = 12

    if historical_analysis_length is not None:
        df_historical = df.tail(historical_analysis_length)
        historical_count = df_historical[column][df_historical[column] == number].count()
        probability_of_occurrence = historical_count / historical_analysis_length
    else:
        probability_of_occurrence = 1 / n_possible

    probabilities = {}
    for k in range(n_future_draws + 1):
        probability = binom.pmf(k, n_future_draws, probability_of_occurrence)
        probabilities[k] = probability

    return probabilities

if __name__ == "__main__":
    # Load your EuroMillions draw history data
    df = pd.read_csv('/content/euromillions-draw-history.csv')

    n_future_draws = 100
    lucky_star_1_numbers_to_analyze = [1, 2, 5, 6]

    print(f"--- Probability of Specific Frequencies in Next {n_future_draws} Draws ---")

    print("\nLucky Star 1 (Theoretical Probability = 1/11):")
    for number in lucky_star_1_numbers_to_analyze:
        prob_ls1_theoretical = probability_specific_lucky_star_frequency(
            df, number, 'Lucky Star 1', n_future_draws
        )
        expected_appearances = round((1/11) * n_future_draws, 2)
        print(f"\nNumber {number}:")
        print(f"  Expected appearances in {n_future_draws} draws: {expected_appearances}")
        print(f"  Probability of appearing exactly {int(expected_appearances)} times: {prob_ls1_theoretical.get(int(expected_appearances), 0):.4f}")
        print(f"  Probability of appearing exactly {int(expected_appearances) - 1} times: {prob_ls1_theoretical.get(int(expected_appearances) - 1, 0):.4f}")
        print(f"  Probability of appearing exactly {int(expected_appearances) + 1} times: {prob_ls1_theoretical.get(int(expected_appearances) + 1, 0):.4f}")

    print("\n--- (Analysis based on historical frequency is also possible by uncommenting below) ---")
    """
    # Example for Lucky Star 1 (using historical frequency from all data)
    print("\nLucky Star 1 (Based on Historical Frequency):")
    all_ls1_counts = df['Lucky Star 1'].value_counts()
    for number in lucky_star_1_numbers_to_analyze:
        historical_count_ls1 = all_ls1_counts.get(number, 0)
        estimated_prob_ls1 = historical_count_ls1 / len(df)
        prob_ls1_historical = probability_specific_lucky_star_frequency(
            df, number, 'Lucky Star 1', n_future_draws, historical_analysis_length=len(df)
        )
        expected_appearances_historical = round(estimated_prob_ls1 * n_future_draws, 2)
        print(f"\nNumber {number}:")
        print(f"  Historical count: {historical_count_ls1}")
        print(f"  Estimated probability of appearing in one draw: {estimated_prob_ls1:.4f}")
        print(f"  Expected appearances in {n_future_draws} draws: {expected_appearances_historical}")
        print(f"  Probability of appearing exactly {int(expected_appearances_historical)} times: {prob_ls1_historical.get(int(expected_appearances_historical), 0):.4f}")
        print(f"  Probability of appearing exactly {int(expected_appearances_historical) - 1} times: {prob_ls1_historical.get(int(expected_appearances_historical) - 1, 0):.4f}")
        print(f"  Probability of appearing exactly {int(expected_appearances_historical) + 1} times: {prob_ls1_historical.get(int(expected_appearances_historical) + 1, 0):.4f}")
    """

--- Probability of Specific Frequencies in Next 100 Draws ---

Lucky Star 1 (Theoretical Probability = 1/11):

Number 1:
  Expected appearances in 100 draws: 9.09
  Probability of appearing exactly 9 times: 0.1380
  Probability of appearing exactly 8 times: 0.1350
  Probability of appearing exactly 10 times: 0.1256

Number 2:
  Expected appearances in 100 draws: 9.09
  Probability of appearing exactly 9 times: 0.1380
  Probability of appearing exactly 8 times: 0.1350
  Probability of appearing exactly 10 times: 0.1256

Number 5:
  Expected appearances in 100 draws: 9.09
  Probability of appearing exactly 9 times: 0.1380
  Probability of appearing exactly 8 times: 0.1350
  Probability of appearing exactly 10 times: 0.1256

Number 6:
  Expected appearances in 100 draws: 9.09
  Probability of appearing exactly 9 times: 0.1380
  Probability of appearing exactly 8 times: 0.1350
  Probability of appearing exactly 10 times: 0.1256

--- (Analysis based on historical frequency is also possible 

# Probability of appearing exactly 9 times: 0.1380: There is approximately a 13.80% chance that the number 2 will be drawn exactly 9 times in the 'Lucky Star 1' position over the next 100 draws.

In [ ]:
def probability_specific_lucky_star_frequency(df, number, column, n_future_draws, historical_analysis_length=None):
    """
    Calculates the probability of a specific Lucky Star number appearing a
    certain number of times in a given number of future draws, based on its
    theoretical probability or historical frequency.

    Args:
        df (pd.DataFrame): The DataFrame containing the draw data.
        number (int): The Lucky Star number to analyze.
        column (str): The Lucky Star column ('Lucky Star 1' or 'Lucky Star 2').
        n_future_draws (int): The number of future draws to consider.
        historical_analysis_length (int, optional): If specified, use the
                                                     frequency from the last
                                                     this many draws to
                                                     estimate probability.
                                                     Defaults to None (uses
                                                     theoretical probability).

    Returns:
        dict: A dictionary containing the probability of the number appearing
              k times (where k ranges from 0 to n_future_draws).
    """
    if column not in ['Lucky Star 1', 'Lucky Star 2']:
        print("Error: Invalid column specified. Use 'Lucky Star 1' or 'Lucky Star 2'.")
        return {}

    if column == 'Lucky Star 1':
        n_possible = 11
    else:
        n_possible = 12

    if historical_analysis_length is not None:
        df_historical = df.tail(historical_analysis_length)
        historical_count = df_historical[column][df_historical[column] == number].count()
        probability_of_occurrence = historical_count / historical_analysis_length
    else:
        probability_of_occurrence = 1 / n_possible

    probabilities = {}
    for k in range(n_future_draws + 1):
        probability = binom.pmf(k, n_future_draws, probability_of_occurrence)
        probabilities[k] = probability

    return probabilities

if __name__ == "__main__":
    # Load your EuroMillions draw history data
    df = pd.read_csv('/content/euromillions-draw-history.csv')

    n_future_draws = 100
    lucky_star_2_numbers_to_analyze = [9, 12, 7, 8, 10, 11]

    print(f"--- Probability of Specific Frequencies in Next {n_future_draws} Draws (Lucky Star 2) ---")

    for number in lucky_star_2_numbers_to_analyze:
        prob_ls2_theoretical = probability_specific_lucky_star_frequency(
            df, number, 'Lucky Star 2', n_future_draws
        )
        expected_appearances = round((1/12) * n_future_draws, 2)
        print(f"\nNumber {number} (Theoretical Probability = 1/12):")
        print(f"  Expected appearances in {n_future_draws} draws: {expected_appearances}")
        print(f"  Probability of appearing exactly {int(expected_appearances)} times: {prob_ls2_theoretical.get(int(expected_appearances), 0):.4f}")
        print(f"  Probability of appearing exactly {int(expected_appearances) - 1} times: {prob_ls2_theoretical.get(int(expected_appearances) - 1, 0):.4f}")
        print(f"  Probability of appearing exactly {int(expected_appearances) + 1} times: {prob_ls2_theoretical.get(int(expected_appearances) + 1, 0):.4f}")

    print("\n--- (Analysis based on historical frequency is also possible by uncommenting below) ---")
    """
    # Example for Lucky Star 2 (using historical frequency from all data)
    print("\nLucky Star 2 (Based on Historical Frequency):")
    all_ls2_counts = df['Lucky Star 2'].value_counts()
    for number in lucky_star_2_numbers_to_analyze:
        historical_count_ls2 = all_ls2_counts.get(number, 0)
        estimated_prob_ls2 = historical_count_ls2 / len(df)
        prob_ls2_historical = probability_specific_lucky_star_frequency(
            df, number, 'Lucky Star 2', n_future_draws, historical_analysis_length=len(df)
        )
        expected_appearances_historical = round(estimated_prob_ls2 * n_future_draws, 2)
        print(f"\nNumber {number}:")
        print(f"  Historical count: {historical_count_ls2}")
        print(f"  Estimated probability of appearing in one draw: {estimated_prob_ls2:.4f}")
        print(f"  Expected appearances in {n_future_draws} draws: {expected_appearances_historical}")
        print(f"  Probability of appearing exactly {int(expected_appearances_historical)} times: {prob_ls2_historical.get(int(expected_appearances_historical), 0):.4f}")
        print(f"  Probability of appearing exactly {int(expected_appearances_historical) - 1} times: {prob_ls2_historical.get(int(expected_appearances_historical) - 1, 0):.4f}")
        print(f"  Probability of appearing exactly {int(expected_appearances_historical) + 1} times: {prob_ls2_historical.get(int(expected_appearances_historical) + 1, 0):.4f}")
    """

--- Probability of Specific Frequencies in Next 100 Draws (Lucky Star 2) ---

Number 9 (Theoretical Probability = 1/12):
  Expected appearances in 100 draws: 8.33
  Probability of appearing exactly 8 times: 0.1445
  Probability of appearing exactly 7 times: 0.1367
  Probability of appearing exactly 9 times: 0.1342

Number 12 (Theoretical Probability = 1/12):
  Expected appearances in 100 draws: 8.33
  Probability of appearing exactly 8 times: 0.1445
  Probability of appearing exactly 7 times: 0.1367
  Probability of appearing exactly 9 times: 0.1342

Number 7 (Theoretical Probability = 1/12):
  Expected appearances in 100 draws: 8.33
  Probability of appearing exactly 8 times: 0.1445
  Probability of appearing exactly 7 times: 0.1367
  Probability of appearing exactly 9 times: 0.1342

Number 8 (Theoretical Probability = 1/12):
  Expected appearances in 100 draws: 8.33
  Probability of appearing exactly 8 times: 0.1445
  Probability of appearing exactly 7 times: 0.1367
  Probability of 

In [ ]:
import pandas as pd
from collections import Counter

def find_most_frequent_main_numbers(df, n=None):
    """
    Finds the most frequent numbers in each of the main ball columns
    (Ball 1 to Ball 5) of a EuroMillions draw history DataFrame.

    Args:
        df (pd.DataFrame): The DataFrame containing the draw data with
                           columns 'Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5'.
        n (int, optional): The number of most frequent numbers to return for
                           each ball column. If None, returns all frequencies.
                           Defaults to None.

    Returns:
        dict: A dictionary where keys are the ball column names and values
              are Counter objects showing the frequency of numbers in that column.
    """
    if not all(col in df.columns for col in ['Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5']):
        print("Error: DataFrame must contain columns 'Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5'.")
        return {}

    frequent_numbers = {}
    ball_columns = ['Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5']

    for col in ball_columns:
        frequent_numbers[col] = Counter(df[col])
        print(f"\nMost Frequent Numbers in {col}:")
        if n is None:
            for number, count in frequent_numbers[col].most_common():
                print(f"  Number {number}: {count} times")
        else:
            for number, count in frequent_numbers[col].most_common(n):
                print(f"  Number {number}: {count} times")

    return frequent_numbers

if __name__ == "__main__":
    # Load your EuroMillions draw history data
    df = pd.read_csv('/content/euromillions-draw-history.csv')

    # Find and print all frequent numbers for each main ball column
    all_frequent_main = find_most_frequent_main_numbers(df)

    # Find and print the top 10 most frequent numbers for each main ball column
    print("\n--- Top 10 Most Frequent Numbers in Each Main Ball Column ---")
    top_10_frequent_main = find_most_frequent_main_numbers(df, n=10)


Most Frequent Numbers in Ball 1:
  Number 1: 6 times
  Number 10: 5 times
  Number 4: 5 times
  Number 7: 5 times
  Number 3: 4 times
  Number 9: 3 times
  Number 8: 3 times
  Number 6: 3 times
  Number 2: 3 times
  Number 5: 2 times
  Number 12: 2 times
  Number 27: 1 times
  Number 13: 1 times
  Number 15: 1 times
  Number 17: 1 times
  Number 22: 1 times
  Number 18: 1 times
  Number 20: 1 times
  Number 19: 1 times
  Number 14: 1 times
  Number 23: 1 times
  Number 16: 1 times

Most Frequent Numbers in Ball 2:
  Number 14: 4 times
  Number 11: 4 times
  Number 15: 4 times
  Number 24: 3 times
  Number 3: 3 times
  Number 17: 3 times
  Number 10: 3 times
  Number 8: 3 times
  Number 33: 3 times
  Number 21: 2 times
  Number 20: 2 times
  Number 22: 2 times
  Number 23: 2 times
  Number 25: 2 times
  Number 13: 2 times
  Number 39: 1 times
  Number 9: 1 times
  Number 36: 1 times
  Number 16: 1 times
  Number 4: 1 times
  Number 30: 1 times
  Number 37: 1 times
  Number 27: 1 times


In [ ]:
def find_most_frequent_main_numbers(df, n=5):
    """
    Finds the top N most frequent numbers in each of the main ball columns
    (Ball 1 to Ball 5) of a EuroMillions draw history DataFrame.

    Args:
        df (pd.DataFrame): The DataFrame containing the draw data with
                           columns 'Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5'.
        n (int): The number of most frequent numbers to return for
                 each ball column. Defaults to 5.

    Returns:
        dict: A dictionary where keys are the ball column names and values
              are lists of the top N (number, count) tuples.
    """
    if not all(col in df.columns for col in ['Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5']):
        print("Error: DataFrame must contain columns 'Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5'.")
        return {}

    top_frequent = {}
    ball_columns = ['Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5']

    for col in ball_columns:
        frequent_numbers = Counter(df[col])
        top_frequent[col] = frequent_numbers.most_common(n)
        print(f"\nTop {n} Most Frequent Numbers in {col}:")
        for number, count in top_frequent[col]:
            print(f"  Number {number}: {count} times")

    return top_frequent

if __name__ == "__main__":
    # Load your EuroMillions draw history data
    df = pd.read_csv('/content/euromillions-draw-history.csv')

    # Find and print the top 5 most frequent numbers for each main ball column
    top_5_frequent_main = find_most_frequent_main_numbers(df, n=5)



Top 5 Most Frequent Numbers in Ball 1:
  Number 1: 6 times
  Number 10: 5 times
  Number 4: 5 times
  Number 7: 5 times
  Number 3: 4 times

Top 5 Most Frequent Numbers in Ball 2:
  Number 14: 4 times
  Number 11: 4 times
  Number 15: 4 times
  Number 24: 3 times
  Number 3: 3 times

Top 5 Most Frequent Numbers in Ball 3:
  Number 29: 4 times
  Number 25: 4 times
  Number 19: 4 times
  Number 41: 3 times
  Number 18: 3 times

Top 5 Most Frequent Numbers in Ball 4:
  Number 42: 5 times
  Number 46: 4 times
  Number 35: 4 times
  Number 29: 4 times
  Number 30: 4 times

Top 5 Most Frequent Numbers in Ball 5:
  Number 50: 8 times
  Number 49: 7 times
  Number 45: 5 times
  Number 47: 5 times
  Number 48: 4 times


In [ ]:
import pandas as pd
from collections import Counter

def find_top_n_most_frequent_all_balls(df, n=5):
    """
    Finds the top N most frequent numbers across all main ball columns
    (Ball 1 to Ball 5) of a EuroMillions draw history DataFrame.

    Args:
        df (pd.DataFrame): The DataFrame containing the draw data with
                           columns 'Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5'.
        n (int): The number of most frequent numbers to return. Defaults to 5.

    Returns:
        list: A list of the top N (number, total_count) tuples, sorted by frequency.
    """
    if not all(col in df.columns for col in ['Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5']):
        print("Error: DataFrame must contain columns 'Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5'.")
        return []

    all_balls_numbers = []
    ball_columns = ['Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5']

    for col in ball_columns:
        all_balls_numbers.extend(df[col].tolist())

    all_balls_counts = Counter(all_balls_numbers)
    top_n_frequent = all_balls_counts.most_common(n)

    print(f"Top {n} Most Frequent Numbers Across All Main Balls:")
    for number, count in top_n_frequent:
        print(f"  Number {number}: {count} times")

    return top_n_frequent

if __name__ == "__main__":
    # Load your EuroMillions draw history data
    df = pd.read_csv('/content/euromillions-draw-history.csv')

    # Find and print the top 5 most frequent numbers across all main balls
    top_5_frequent_all = find_top_n_most_frequent_all_balls(df, n=5)

Top 5 Most Frequent Numbers Across All Main Balls:
  Number 29: 9 times
  Number 10: 8 times
  Number 33: 8 times
  Number 42: 8 times
  Number 50: 8 times


#   rows  were   at least 3 0f 5  appear  in one row

In [ ]:
import pandas as pd

def find_rows_with_at_least_n_matches(df, target_numbers, n=3):
    """
    Finds rows in a EuroMillions draw history DataFrame where at least N
    of the target numbers appear in the main ball columns (Ball 1 to Ball 5).

    Args:
        df (pd.DataFrame): The DataFrame containing the draw data with
                           columns 'Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5'.
        target_numbers (list): A list of numbers to check for in each row.
        n (int): The minimum number of target numbers that must appear in a row.
                 Defaults to 3.

    Returns:
        pd.DataFrame: A DataFrame containing the rows that meet the criteria.
                      Returns an empty DataFrame if no rows match.
    """
    if not all(col in df.columns for col in ['Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5']):
        print("Error: DataFrame must contain columns 'Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5'.")
        return pd.DataFrame()

    matching_rows = []
    ball_columns = ['Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5']

    for index, row in df.iterrows():
        drawn_numbers = row[ball_columns].tolist()
        matches = 0
        for number in target_numbers:
            if number in drawn_numbers:
                matches += 1
        if matches >= n:
            matching_rows.append(row)

    return pd.DataFrame(matching_rows)

if __name__ == "__main__":
    # Load your EuroMillions draw history data
    df = pd.read_csv('/content/euromillions-draw-history.csv')

    # The top 5 most frequent numbers you provided
    top_5_frequent = [29, 10, 33, 42, 50]

    # Find rows where at least 3 of these numbers appear
    matching_df = find_rows_with_at_least_n_matches(df, top_5_frequent, n=3)

    if not matching_df.empty:
        print(f"Rows where at least 3 of the numbers {top_5_frequent} appeared:")
        print(matching_df)
    else:
        print(f"No rows found where at least 3 of the numbers {top_5_frequent} appeared.")

Rows where at least 3 of the numbers [29, 10, 33, 42, 50] appeared:
       DrawDate  Ball 1  Ball 2  Ball 3  Ball 4  Ball 5  Lucky Star 1  \
30  20-Dec-2024      10      14      21      33      50             6   

    Lucky Star 2 UK Millionaire Maker  European Millionaire Maker  DrawNumber  
30             9            MBPP66710                         NaN        1800  


In [ ]:
import pandas as pd

def find_rows_with_at_least_n_matches(df, target_numbers, n=4):
    """
    Finds rows in a EuroMillions draw history DataFrame where at least N
    of the target numbers appear in the main ball columns (Ball 1 to Ball 5).

    Args:
        df (pd.DataFrame): The DataFrame containing the draw data with
                           columns 'Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5'.
        target_numbers (list): A list of numbers to check for in each row.
        n (int): The minimum number of target numbers that must appear in a row.
                 Defaults to 4.

    Returns:
        pd.DataFrame: A DataFrame containing the rows that meet the criteria.
                      Returns an empty DataFrame if no rows match.
    """
    if not all(col in df.columns for col in ['Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5']):
        print("Error: DataFrame must contain columns 'Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5'.")
        return pd.DataFrame()

    matching_rows = []
    ball_columns = ['Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5']

    for index, row in df.iterrows():
        drawn_numbers = row[ball_columns].tolist()
        matches = 0
        for number in target_numbers:
            if number in drawn_numbers:
                matches += 1
        if matches >= n:
            matching_rows.append(row)

    return pd.DataFrame(matching_rows)

if __name__ == "__main__":
    # Load your EuroMillions draw history data
    df = pd.read_csv('/content/euromillions-draw-history.csv')

    # The top 5 most frequent numbers you provided
    top_5_frequent = [29, 10, 33, 42, 50]

    # Find rows where at least 4 of these numbers appear
    matching_df = find_rows_with_at_least_n_matches(df, top_5_frequent, n=4)

    if not matching_df.empty:
        print(f"Rows where at least 4 of the numbers {top_5_frequent} appeared:")
        print(matching_df)
    else:
        print(f"No rows found where at least 4 of the numbers {top_5_frequent} appeared.")

No rows found where at least 4 of the numbers [29, 10, 33, 42, 50] appeared.


find rowds      matching_rows = []

ball_columns = ['Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5'] were [29, 10, 33, 42, 50] at least 3 apper in a Ball columns , and  ['Lucky Star 1'] 1, 2, 5, or 6 drawn_ls2 = row['Lucky Star 2'] one of [9 12 7 8 10 11

In [ ]:
import pandas as pd

def find_rows_with_ball_and_ls(df, target_balls, ls1_targets, ls2_targets, min_ball_matches=3):
    """
    Finds rows where at least a certain number of target balls appear AND
    Lucky Star 1 matches one of the LS1 targets OR Lucky Star 2 matches one
    of the LS2 targets.

    Args:
        df (pd.DataFrame): The DataFrame containing the draw data.
        target_balls (list): A list of target main ball numbers.
        ls1_targets (list): A list of target Lucky Star 1 numbers.
        ls2_targets (list): A list of target Lucky Star 2 numbers.
        min_ball_matches (int): The minimum number of target balls that must appear.
                                Defaults to 3.

    Returns:
        pd.DataFrame: A DataFrame containing the matching rows.
    """
    if not all(col in df.columns for col in ['Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5', 'Lucky Star 1', 'Lucky Star 2']):
        print("Error: DataFrame must contain all required columns.")
        return pd.DataFrame()

    matching_rows = []
    ball_columns = ['Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5']

    for index, row in df.iterrows():
        drawn_balls = row[ball_columns].tolist()
        drawn_ls1 = row['Lucky Star 1']
        drawn_ls2 = row['Lucky Star 2']

        ball_matches = 0
        for ball in target_balls:
            if ball in drawn_balls:
                ball_matches += 1

        ls1_match = drawn_ls1 in ls1_targets
        ls2_match = drawn_ls2 in ls2_targets

        if ball_matches >= min_ball_matches and (ls1_match or ls2_match):
            matching_rows.append(row)

    return pd.DataFrame(matching_rows)

if __name__ == "__main__":
    # Load your EuroMillions draw history data
    df = pd.read_csv('/content/euromillions-draw-history.csv')

    # Target main ball numbers
    target_main_balls = [29, 10, 33, 42, 50]
    min_matches = 3

    # Target Lucky Star numbers
    target_ls1 = [1, 2, 5, 6]
    target_ls2 = [9, 12, 7, 8, 10, 11]

    # Find rows that meet the criteria
    matching_df = find_rows_with_ball_and_ls(df, target_main_balls, target_ls1, target_ls2, min_matches)

    if not matching_df.empty:
        print(f"Rows with at least {min_matches} of the balls {target_main_balls} and")
        print(f"Lucky Star 1 in {target_ls1} OR Lucky Star 2 in {target_ls2}:")
        print(matching_df)
    else:
        print(f"No rows found with at least {min_matches} of the balls {target_main_balls} and")
        print(f"Lucky Star 1 in {target_ls1} OR Lucky Star 2 in {target_ls2}.")

Rows with at least 3 of the balls [29, 10, 33, 42, 50] and
Lucky Star 1 in [1, 2, 5, 6] OR Lucky Star 2 in [9, 12, 7, 8, 10, 11]:
       DrawDate  Ball 1  Ball 2  Ball 3  Ball 4  Ball 5  Lucky Star 1  \
30  20-Dec-2024      10      14      21      33      50             6   

    Lucky Star 2 UK Millionaire Maker  European Millionaire Maker  DrawNumber  
30             9            MBPP66710                         NaN        1800  


In [ ]:
from math import comb

def probability_event():
    """
    Calculates the probability of the specified EuroMillions event occurring
    in a single future draw.

    Event:
    - At least 3 of the main balls are from the set [29, 10, 33, 42, 50].
    - AND (Lucky Star 1 is one of [1, 2, 5, 6] OR Lucky Star 2 is one of [9, 12, 7, 8, 10, 11]).
    """
    # Probability of at least 3 target main balls
    total_main_combinations = comb(50, 5)
    target_balls = 5
    non_target_balls = 45

    prob_3_matches = (comb(target_balls, 3) * comb(non_target_balls, 2)) / total_main_combinations
    prob_4_matches = (comb(target_balls, 4) * comb(non_target_balls, 1)) / total_main_combinations
    prob_5_matches = (comb(target_balls, 5) * comb(non_target_balls, 0)) / total_main_combinations

    prob_at_least_3_target_balls = prob_3_matches + prob_4_matches + prob_5_matches

    # Probability of the Lucky Star condition
    ls1_target = 4
    ls2_target = 6
    total_ls = 12

    prob_ls1_match = ls1_target / total_ls
    prob_ls2_match = ls2_target / total_ls
    prob_both_match = prob_ls1_match * prob_ls2_match  # Assuming independence

    prob_ls1_or_ls2_match = prob_ls1_match + prob_ls2_match - prob_both_match

    # Probability of both events occurring (assuming independence)
    probability = prob_at_least_3_target_balls * prob_ls1_or_ls2_match

    return probability

if __name__ == "__main__":
    probability = probability_event()
    print(f"The probability of the specified event in a single EuroMillions draw is approximately: {probability:.8f}")

The probability of the specified event in a single EuroMillions draw is approximately: 0.00318614


# At least 3 of the 5 main balls drawn will be from the set: [29, 10, 33, 42, 50].

This includes the possibilities of exactly 3, exactly 4, or all 5 of these numbers being drawn as main balls.
AND

Either the first Lucky Star drawn will be one of the numbers in the set: [1, 2, 5, 6], OR the second Lucky Star drawn will be one of the numbers in the set: [9, 12, 7, 8, 10, 11], OR both of these conditions are true.

So, for every 1000 EuroMillions draws, you might expect this specific combination of events to occur roughly 3 to 4 times, on average.

#    cost of calculated probability o.03 for 1000  drow

In [ ]:
from math import comb

def probability_event():
    """
    Calculates the probability of the specified EuroMillions event occurring
    in a single future draw.

    Event:
    - At least 3 of the main balls are from the set [29, 10, 33, 42, 50].
    - AND (Lucky Star 1 is one of [1, 2, 5, 6] OR Lucky Star 2 is one of [9, 12, 7, 8, 10, 11]).
    """
    # Probability of at least 3 target main balls
    total_main_combinations = comb(50, 5)
    target_balls = 5
    non_target_balls = 45

    prob_3_matches = (comb(target_balls, 3) * comb(non_target_balls, 2)) / total_main_combinations
    prob_4_matches = (comb(target_balls, 4) * comb(non_target_balls, 1)) / total_main_combinations
    prob_5_matches = (comb(target_balls, 5) * comb(non_target_balls, 0)) / total_main_combinations

    prob_at_least_3_target_balls = prob_3_matches + prob_4_matches + prob_5_matches

    # Probability of the Lucky Star condition
    ls1_target = 4
    ls2_target = 6
    total_ls = 12

    prob_ls1_match = ls1_target / total_ls
    prob_ls2_match = ls2_target / total_ls
    prob_both_match = prob_ls1_match * prob_ls2_match  # Assuming independence

    prob_ls1_or_ls2_match = prob_ls1_match + prob_ls2_match - prob_both_match

    # Probability of both events occurring (assuming independence)
    probability = prob_at_least_3_target_balls * prob_ls1_or_ls2_match

    return probability

def calculate_expected_cost_per_event(probability_per_draw, num_draws, cost_per_ticket):
    """
    Calculates the expected cost associated with the occurrence of a specific
    event over a number of draws. This is a theoretical average.

    Args:
        probability_per_draw (float): The probability of the event in a single draw.
        num_draws (int): The number of draws considered.
        cost_per_ticket (float): The cost of one EuroMillions ticket.

    Returns:
        float: The expected cost per occurrence of the event. Returns infinity
               if the expected number of occurrences is zero.
    """
    expected_occurrences = probability_per_draw * num_draws
    if expected_occurrences > 0:
        total_cost = cost_per_ticket * num_draws
        expected_cost_per_event = total_cost / expected_occurrences
        return expected_cost_per_event
    else:
        return float('inf')

if __name__ == "__main__":
    probability_single_draw = probability_event()
    num_draws = 100
    cost_per_ticket = 2.50  # Price in Ireland

    expected_cost = calculate_expected_cost_per_event(probability_single_draw, num_draws, cost_per_ticket)

    print(f"The probability of the specified event in a single EuroMillions draw is approximately: {probability_single_draw:.8f}")
    print(f"Considering {num_draws} draws with a ticket price of €{cost_per_ticket:.2f}:")
    if expected_cost == float('inf'):
        print("The expected number of occurrences of this event is zero over 100 draws.")
    else:
        print(f"The expected cost associated with each occurrence of this event is approximately: €{expected_cost:.2f}")

    print("\nImportant Note:")
    print("This is a theoretical calculation based on probability.")
    print("The lottery is random, and actual outcomes can vary significantly.")
    print("This does not guarantee that you will spend this exact amount to see this event occur.")
    print("Play responsibly.")

The probability of the specified event in a single EuroMillions draw is approximately: 0.00318614
Considering 100 draws with a ticket price of €2.50:
The expected cost associated with each occurrence of this event is approximately: €784.65

Important Note:
This is a theoretical calculation based on probability.
The lottery is random, and actual outcomes can vary significantly.
This does not guarantee that you will spend this exact amount to see this event occur.
Play responsibly.


In [ ]:
from collections import Counter
from itertools import combinations

def get_top_frequent_numbers_per_ball(df, n=5):
    """
    Finds the top N most frequent numbers in each of the main ball columns.

    Args:
        df (pd.DataFrame): The DataFrame containing the draw data.
        n (int): The number of top frequent numbers to return.

    Returns:
        dict: A dictionary where keys are ball column names and values are
              lists of the top N most frequent numbers.
    """
    top_frequent = {}
    ball_columns = ['Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5']
    for col in ball_columns:
        counts = Counter(df[col])
        top_n = [num for num, count in counts.most_common(n)]
        top_frequent[col] = top_n
    return top_frequent

def generate_potential_combinations(top_frequent):
    """
    Generates a few potential 5-number combinations based on the top
    frequent numbers from each ball position. This is NOT a statistically
    sound prediction method.

    Args:
        top_frequent (dict): A dictionary of top frequent numbers per ball.

    Returns:
        list: A list of 5 potential main number combinations.
    """
    combinations_list = []

    # Try combining the top number from each ball (if unique)
    top_from_each = []
    seen = set()
    for col in ['Ball 1', 'Ball 2', 'Ball 3', 'Ball 4', 'Ball 5']:
        if top_frequent[col] and top_frequent[col][0] not in seen:
            top_from_each.append(top_frequent[col][0])
            seen.add(top_frequent[col][0])

    if len(top_from_each) >= 5:
        combinations_list.append(tuple(sorted(top_from_each[:5])))

    # Try other combinations using the top 2-3 from each ball (avoiding repeats)
    all_top = []
    for col in top_frequent:
        all_top.extend(top_frequent[col][:3])
    unique_top = sorted(list(set(all_top)))

    # Generate a few combinations (this is a very basic approach)
    for combo in combinations(unique_top, 5):
        if len(combinations_list) < 5:
            combinations_list.append(combo)
        else:
            break

    return combinations_list

if __name__ == "__main__":
    import pandas as pd

    # Load your EuroMillions draw history data
    try:
        df = pd.read_csv('/content/euromillions-draw-history.csv')
    except FileNotFoundError:
        print("Error: EuroMillions draw history CSV file not found at /content/")
        exit()

    top_frequent_numbers = get_top_frequent_numbers_per_ball(df)
    print("Top 5 Most Frequent Numbers in Each Ball Position:")
    for ball, numbers in top_frequent_numbers.items():
        print(f"  {ball}: {numbers}")

    potential_main_combinations = generate_potential_combinations(top_frequent_numbers)

    lucky_star_1_targets = [1, 2, 5, 6]
    lucky_star_2_targets = [9, 12, 7, 8, 10, 11]

    print("\n5 Potential Combinations Based on Frequent Numbers (NOT a prediction):")
    for i, combo in enumerate(potential_main_combinations):
        print(f"  Combination {i+1}: {sorted(list(combo))}")

    print("\nPossible Combinations for Prize Tiers (Based on Frequent Numbers and Lucky Star Sets):")

    print("\n5 + 1 (5 Main Numbers + 1 Lucky Star):")
    for combo in potential_main_combinations:
        for ls1 in lucky_star_1_targets[:min(len(lucky_star_1_targets), 2)]:
            print(f"  {sorted(list(combo))} + LS1: {ls1}")

    print("\n5 (5 Main Numbers):")
    for combo in potential_main_combinations:
        print(f"  {sorted(list(combo))} + (Any two Lucky Stars)") # Lucky Stars don't matter for this tier

    print("\n5 + 2 (5 Main Numbers + 2 Lucky Stars):")
    for combo in potential_main_combinations:
        for ls1 in lucky_star_1_targets[:min(len(lucky_star_1_targets), 2)]:
            for ls2 in lucky_star_2_targets[:min(len(lucky_star_2_targets), 2)]:
                if ls1 != ls2:
                    print(f"  {sorted(list(combo))} + LS1: {ls1}, LS2: {ls2}")

    print("\nImportant Note:")
    print("This code generates combinations based on historical frequency, which is NOT a reliable predictor of future lottery outcomes.")
    print("EuroMillions draws are random, and every number has an equal chance of being drawn in each draw.")
    print("Play responsibly and understand that lottery results are based on chance.")

Top 5 Most Frequent Numbers in Each Ball Position:
  Ball 1: [1, 10, 4, 7, 3]
  Ball 2: [14, 11, 15, 24, 3]
  Ball 3: [29, 25, 19, 41, 18]
  Ball 4: [42, 46, 35, 29, 30]
  Ball 5: [50, 49, 45, 47, 48]

5 Potential Combinations Based on Frequent Numbers (NOT a prediction):
  Combination 1: [1, 14, 29, 42, 50]
  Combination 2: [1, 4, 10, 11, 14]
  Combination 3: [1, 4, 10, 11, 15]
  Combination 4: [1, 4, 10, 11, 19]
  Combination 5: [1, 4, 10, 11, 25]

Possible Combinations for Prize Tiers (Based on Frequent Numbers and Lucky Star Sets):

5 + 1 (5 Main Numbers + 1 Lucky Star):
  [1, 14, 29, 42, 50] + LS1: 1
  [1, 14, 29, 42, 50] + LS1: 2
  [1, 4, 10, 11, 14] + LS1: 1
  [1, 4, 10, 11, 14] + LS1: 2
  [1, 4, 10, 11, 15] + LS1: 1
  [1, 4, 10, 11, 15] + LS1: 2
  [1, 4, 10, 11, 19] + LS1: 1
  [1, 4, 10, 11, 19] + LS1: 2
  [1, 4, 10, 11, 25] + LS1: 1
  [1, 4, 10, 11, 25] + LS1: 2

5 (5 Main Numbers):
  [1, 14, 29, 42, 50] + (Any two Lucky Stars)
  [1, 4, 10, 11, 14] + (Any two Lucky Stars)
  [1

In [ ]:
import pandas as pd

def check_combinations_in_draws(df, combinations_to_check):
    """
    Checks if specific main number and Lucky Star combinations are present
    in the draw history.

    Args:
        df (pd.DataFrame): The DataFrame containing the draw data.
        combinations_to_check (list): A list of tuples, where each tuple is:
                                      (main_numbers_list, lucky_star_1, lucky_star_2).

    Returns:
        dict: A dictionary where keys are the combinations and values are boolean
              (True if found, False otherwise).
    """
    results = {}
    for main_numbers, ls1, ls2 in combinations_to_check:
        found = False
        sorted_main = sorted(main_numbers)
        for _, row in df.iterrows():
            draw_main = sorted([row['Ball 1'], row['Ball 2'], row['Ball 3'], row['Ball 4'], row['Ball 5']])
            draw_ls1 = row['Lucky Star 1']
            draw_ls2 = row['Lucky Star 2']
            if draw_main == sorted_main and draw_ls1 == ls1 and draw_ls2 == ls2:
                found = True
                break
        results[(tuple(main_numbers), ls1, ls2)] = found
    return results

if __name__ == "__main__":
    # Load your EuroMillions draw history data
    try:
        df = pd.read_csv('/content/euromillions-draw-history.csv')
    except FileNotFoundError:
        print("Error: EuroMillions draw history CSV file not found at /content/")
        exit()

    combinations_to_check = [
        ([1, 14, 29, 42, 50], 1, 9),
        ([1, 14, 29, 42, 50], 1, 12),
        ([1, 14, 29, 42, 50], 2, 9),
        ([1, 14, 29, 42, 50], 2, 12),
        ([1, 4, 10, 11, 14], 1, 9),
        ([1, 4, 10, 11, 14], 1, 12),
        ([1, 4, 10, 11, 14], 2, 9),
        ([1, 4, 10, 11, 14], 2, 12),
        ([1, 4, 10, 11, 15], 1, 9),
        ([1, 4, 10, 11, 15], 1, 12),
        ([1, 4, 10, 11, 15], 2, 9),
        ([1, 4, 10, 11, 15], 2, 12),
        ([1, 4, 10, 11, 19], 1, 9),
        ([1, 4, 10, 11, 19], 1, 12),
        ([1, 4, 10, 11, 19], 2, 9),
        ([1, 4, 10, 11, 19], 2, 12),
        ([1, 4, 10, 11, 25], 1, 9),
        ([1, 4, 10, 11, 25], 1, 12),
        ([1, 4, 10, 11, 25], 2, 9),
        ([1, 4, 10, 11, 25], 2, 12),
    ]

    results = check_combinations_in_draws(df, combinations_to_check)

    print("Checking for the presence of specific combinations in the draw history:")
    for combo, present in results.items():
        main_numbers, ls1, ls2 = combo
        print(f"Main Numbers: {sorted(list(main_numbers))}, LS1: {ls1}, LS2: {ls2} - Present: {present}")

Checking for the presence of specific combinations in the draw history:
Main Numbers: [1, 14, 29, 42, 50], LS1: 1, LS2: 9 - Present: False
Main Numbers: [1, 14, 29, 42, 50], LS1: 1, LS2: 12 - Present: False
Main Numbers: [1, 14, 29, 42, 50], LS1: 2, LS2: 9 - Present: False
Main Numbers: [1, 14, 29, 42, 50], LS1: 2, LS2: 12 - Present: False
Main Numbers: [1, 4, 10, 11, 14], LS1: 1, LS2: 9 - Present: False
Main Numbers: [1, 4, 10, 11, 14], LS1: 1, LS2: 12 - Present: False
Main Numbers: [1, 4, 10, 11, 14], LS1: 2, LS2: 9 - Present: False
Main Numbers: [1, 4, 10, 11, 14], LS1: 2, LS2: 12 - Present: False
Main Numbers: [1, 4, 10, 11, 15], LS1: 1, LS2: 9 - Present: False
Main Numbers: [1, 4, 10, 11, 15], LS1: 1, LS2: 12 - Present: False
Main Numbers: [1, 4, 10, 11, 15], LS1: 2, LS2: 9 - Present: False
Main Numbers: [1, 4, 10, 11, 15], LS1: 2, LS2: 12 - Present: False
Main Numbers: [1, 4, 10, 11, 19], LS1: 1, LS2: 9 - Present: False
Main Numbers: [1, 4, 10, 11, 19], LS1: 1, LS2: 12 - Present: